Replace existing machine checker with new operator controller Machine

cadenmarchese opened the pull request:
### Which issue this PR addresses:

User Story 8749435. Removes `machinechecker.go`, adds a new operator controller `Machine`.

<!--
Please include a link to the ADO work item as well as any GitHub issues.

Usage: `Fixes #<GitHub issue number>`, or `Fixes (paste link of issue)`.
-->

### What this PR does / why we need it:

We need to ensure customers have at least two worker nodes (or three) to ensure that the cluster is supported,  upgrades are able to complete, and the OpenShift operators are not degraded. As detailed in the user story, the first step to this is to remove the `machinechecker.go` logic from the `Checker` controller and add it into a new controller `Machine`. This new checker validates VMs and outputs a status to the cluster object.

The next step will be to PR a new controller `MachineSet` that will scale up worker MachineSets if it detects less than a minumum of two (or three) worker nodes.

<!--
Include a brief summary of what the PR is intended to accomplish and how the PR
does it. (2-3 sentences)
-->

### Test plan for issue:

Steps taken to test:
Ran the operator master locally with kubeconfig, and confirmed that the new controller was loaded.
Ensured that a status was outputted to the cluster object.
Ensured that the code passed `make test-go`.
Ran test in `machine_controller_test.go`.

<!--
How did you test that this PR works?

- Are there unit tests?
- Are there integration/e2e tests?
- If it is not possible to write automated tests, explain why and document how
  to manually test and verify the feature.
-->

### Is there any documentation that needs to be updated for this PR?

No, however we may need to update the [operator E2E tests](https://github.com/Azure/ARO-RP/tree/master/pkg/operator#how-to-run-operator-e2e-test).

<!--
- If yes and the docs are in GitHub, include doc updates in the PR.
- If yes and the docs are not in GitHub (i.e. ADO wiki), include a link to the
  docs.
- If no, explain why (e.g. "tech debt cleanup, N/A").
-->

troy0820 commented on pkg/operator/controllers/checker/checker_controller.go:
```diff
@@ -38,7 +38,7 @@ func NewReconciler(log *logrus.Entry, maocli maoclient.Interface, arocli aroclie
 	if role == operator.RoleMaster {
 		checkers = append(checkers,
-			NewMachineChecker(log, maocli, arocli, role, isLocalDevelopmentMode),
+			// NewMachineChecker(log, maocli, arocli, role, isLocalDevelopmentMode),
```
```suggestion
		
```
This doesn't need to be here anymore because you moved this out of the checker_controller. You can delete this.

cadenmarchese replied:
Fixed, resolving.

troy0820 commented on pkg/operator/controllers/machine/const.go:
```diff
@@ -0,0 +1,9 @@
+
+const (
+	// replaces pkg/operator/controllers/checker/const.go
+	machineSetsNamespace = "openshift-machine-api"
```
@cadenmarchese your lint-go tests are failing because you need to delete machineSetsNamespace in `pkg/operator/controllers/checker/const.go`.  It's not being used anymore because it is now here.

cadenmarchese replied:
Fixed, thanks!

troy0820 commented on pkg/operator/controllers/machine/machine_controller.go:
```diff
@@ -0,0 +1,77 @@
+		r.log.Errorf("%d", err)
+	}
+
+	return reconcile.Result{RequeueAfter: time.Hour, Requeue: true}, err
```
Still not sure if we want to do this once an hour or because we split this into its own controller for it to reconcile on its own by what it watches.  @mjudeikis 
Also, I believe you may want to get the cluster object before you do the `r.Check` when reconciling.  
`instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, request.Name, metav1.GetOptions{})`

cadenmarchese replied:
> 
> Also, I believe you may want to get the cluster object before you do the r.Check when reconciling.
> `instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, request.Name, metav1.GetOptions{})`
> 
>

Do we want the reconciler to do anything with the cluster object, or just ensure that it can be loaded in?

mjudeikis replied:
is there something in the cluster object we need? If no, we should not be getting it here. 
Whats needs to change is that timer reconcile needs to be removed and we should reconcile each time Machine object changes (look how other controller enqueue on object changes)

cadenmarchese replied:
I've removed the time-based requeue in the reconciler, and tested to ensure that cluster object status is getting updated each time there's a change.

jewzaam commented on pkg/operator/controllers/machine/machine.go:
```diff
@@ -75,10 +55,12 @@ func (r *MachineChecker) machineValid(ctx context.Context, machine *machinev1bet
 		errs = append(errs, fmt.Errorf("machine %s: invalid VM size '%s'", machine.Name, machineProviderSpec.VMSize))
 	}
 
+	// 	Validate disk size in machine provider spec
```
```suggestion
	// Validate disk size in machine provider spec
```

mjudeikis commented on go.mod:
```diff
@@ -111,7 +111,7 @@ require (
 	sigs.k8s.io/cluster-api-provider-azure v0.4.12
 	sigs.k8s.io/cluster-api-provider-openstack v0.3.3 // indirect
-	sigs.k8s.io/controller-runtime v0.8.2
+	sigs.k8s.io/controller-runtime v0.8.3
```
Normally we dont change things like this if there is no requirement :)

mjudeikis commented on pkg/operator/controllers/machine/machine_controller.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+func NewMachineReconciler(log *logrus.Entry, maocli maoclient.Interface, arocli aroclient.Interface, isLocalDevelopmentMode bool, role string) *MachineReconciler {
+
```
drop the space here :)

mjudeikis commented on pkg/operator/controllers/machine/machine_controller.go:
```diff
@@ -0,0 +1,77 @@
+	// TODO(mj): Reconcile will eventually be receiving a ctx (https://github.com/kubernetes-sigs/controller-runtime/blob/7ef2da0bc161d823f084ad21ff5f9c9bd6b0cc39/pkg/reconcile/reconcile.go#L93)
+	ctx := context.TODO()
+
+	err := r.Check(ctx)
+	if err != nil {
+		r.log.Errorf("%d", err)
+	}
```
Drop this at all and bring function content here.

mjudeikis reviewed (changes requested):
Overall close :)

What thing you might want to try/test here is commit management,
We try to split commit into re-viewable entities. Examples here could be:
1 Commit for vendor
2. no-op checked to dedicated controller
3. check() merge with reconcile

When im done with code and have tone of commits (like you do now), I do soft reset to top and  redo code with nice commit messages.

If this does not work for this PR - just squash everything to one commit as its quite small change. But in general its good practice to keep commits isolated logically.

cadenmarchese commented:
The recently added azurensg operator controller adds a const.go file that is identical to the one in this PR. I believe that may be why lint is failing after rebase, but tests are passing on my end. Let me know the preferred way to resolve this.

nilsanderselde commented on pkg/operator/controllers/checker/const.go:
```diff
@@ -4,5 +4,6 @@ package checker
 const (
-	machineSetsNamespace = "openshift-machine-api"
+	azureCredentialSecretNamespace = "kube-system"
+	azureCredentialSecretName      = "azure-credentials"
```
@cadenmarchese these constants are not used anywhere, which is why the go linting failed. they should be removed, or used somewhere if they are needed.

```
pkg/operator/controllers/checker/const.go:7:2: `azureCredentialSecretNamespace` is unused (deadcode)
	azureCredentialSecretNamespace = "kube-system"
	^
pkg/operator/controllers/checker/const.go:8:2: `azureCredentialSecretName` is unused (deadcode)
	azureCredentialSecretName      = "azure-credentials"
	^
```

if you run `make lint-go` locally it will fail there as well.

nilsanderselde replied:
You can delete this file.

cadenmarchese replied:
Thanks!

m1kola commented on pkg/operator/controllers/checker/checker_controller.go:
```diff
@@ -37,7 +37,6 @@ func NewReconciler(log *logrus.Entry, maocli maoclient.Interface, arocli aroclie
-			NewMachineChecker(log, maocli, arocli, role, isLocalDevelopmentMode),
 			NewServicePrincipalChecker(log, maocli, arocli, kubernetescli, role),
 		)
 	}
```
`SetupWithManager` needs to be updated here: we no longer want to watch `machinev1beta1.Machine`, it seems

m1kola replied:
Hm. Or do we still need it for some other checker?

cadenmarchese replied:
We could take it or leave it, but I think that if changes are made to the machine objects we would still want to run `Checker` for internet and service principal checks. Let me know what you think.

mjudeikis replied:
These should be covered by re-queuing. I agree with @m1kola, lets remove them from `checker`

m1kola replied:
Bump.

cadenmarchese replied:
I've removed that watch.

m1kola commented on pkg/operator/controllers/checker/checker_controller.go:
```diff
@@ -37,7 +37,6 @@ func NewReconciler(log *logrus.Entry, maocli maoclient.Interface, arocli aroclie
 
 	if role == operator.RoleMaster {
 		checkers = append(checkers,
-			NewMachineChecker(log, maocli, arocli, role, isLocalDevelopmentMode),
```
It looks like we no longer need `isLocalDevelopmentMode` argument in `NewReconciler`.

m1kola commented on pkg/operator/controllers/machine/machine.go:
```diff
@@ -75,37 +51,41 @@ func (r *MachineChecker) machineValid(ctx context.Context, machine *machinev1bet
 		errs = append(errs, fmt.Errorf("machine %s: invalid image '%v'", machine.Name, machineProviderSpec.Image))
 	}
 
+	// Validate managed identity
```
I appreciate the effort of documenting the codebase, but some of these comments are unnecessary. Like this one: the code is straightforward and probably doesn't diserve a comment.

nilsanderselde replied:
I disagree. One-line comments do not negatively impact anyone's ability to read the code itself, if that is what one prefers to do. On the other hand, the absence of such comments can make it take much longer to understand 1. what is happening, and 2. what is intended to happen (they can differ, so a comment can help make that clear).

I would like to see more such comments in the codebase, so that it is easier for more people to jump in and understand enough to make contributions.

cadenmarchese replied:
I appreciate comments such as this when I'm skimming over the code, but if it's causing clutter we can remove the shorter ones.

m1kola replied:
I'm very much in favour of having comments in the codebase. What I'm trying to say is - let's avoid code and comments like this:

```go
// Add two values
foo := baz + bar
```

I'm giving this as an example becuase error is saying "machine %s: invalid managedIdentity" and implies that the check has something to do with managedIdentity validation.

> 2. what is intended to happen (they can differ, so a comment can help make that clear).

For this we should use tests :)

cadenmarchese replied:
I agree, I've removed some unnecessary comments (ones with good error messages).

m1kola commented on pkg/operator/controllers/machine/machine_controller.go:
```diff
@@ -0,0 +1,94 @@
+	})
+
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(&arov1alpha1.Cluster{}, builder.WithPredicates(aroClusterPredicate))
```
I think we do not need to watch for this object: we are only watching for machine changes, right?

I imagine it should be something simply as


```go

        func (r *MachineReconciler) SetupWithManager(mgr ctrl.Manager) error {
        	return ctrl.NewControllerManagedBy(mgr).
        		For(machinev1beta1.Machine{}).
        		Named(controllers.MachineControllerName).
        		Complete(r)
        }
```

m1kola replied:
But probably with `if r.role == operator.RoleMaster`

cadenmarchese replied:
I've realized that since we are breaking away the former `machinechecker.go` from the `Checker` controller, we should no longer need `if r.role == operator.RoleMaster` since that logic is now in `operator.go.` The SetupWithManager() has been simplified according to your suggestions.

m1kola replied:
@cadenmarchese hm... Are you sure about `if r.role == operator.RoleMaster`? Without this we will be running two reconcilers: one on one of the master nodes and another on one of the worker nodes. There (probably) will be a rance condition between two of them on an attempt to update status of the object. It will not manifest loudly due to retries, but if we can avoid it - we should.

Or do I miss something?

cadenmarchese replied:
Agreed, and now that we have that logic in operator.go, we shouldn't need it here - machine controller will not run if operator role is worker: https://github.com/Azure/ARO-RP/blob/0b81a3f4c81b0b3aa0b5155e53676e12ba761e98/cmd/aro/operator.go#L99 https://github.com/Azure/ARO-RP/blob/0b81a3f4c81b0b3aa0b5155e53676e12ba761e98/cmd/aro/operator.go#L166-L170

m1kola commented on pkg/operator/controllers/machine/machine_controller_test.go:
```diff
@@ -176,7 +176,7 @@ func TestMachineValid(t *testing.T) {
 			errs := r.machineValid(ctx, tt.machine, isMaster)
```
Because it is a proper `Reconciler` now we should test it via `Reconcile` method. Otherwise we leave out some code without coverage.

cadenmarchese replied:
Agreed, I'm working on improving the unit test and will follow up.

cadenmarchese replied:
@m1kola @jewzaam @mjudeikis the original `machinechecker_test` was only meant to check machineValid, and to account for all test cases here would add quite a bit more to this PR. Additionally, many operator controllers don't have unit tests at all, and so I'm wondering if the unit test improvements proposed here should instead go into a new user story for improving/adding unit tests across all operator controllers.

mjudeikis replied:
I think what @m1kola is saying that we should be Testing `Reconcile` instead of `machineValid` . So for now lets just rewrite same tests for to be testing using Reconcile (See monitoring controller for example). And if you could raise story to improve testing in general as follow-up

m1kola replied:
I think we should improve this test while we are refactoring this codebase. We can deal with others later (as MJ suggested - please raise a story for this).

> to account for all test cases here would add quite a bit more to this PR.

Yes, it will be a bit of extra work, but you will not have to cover extra cases here. You will need to change how you instantiate `MachineReconciler` in the test, add some mocks (for `r.arocli` at least) and replace test assertion `if !reflect.DeepEqual(errs, tt.wantErrs)` which a check that looks into actual conditions on the cluster object.

Let me know if you need help with this.

cadenmarchese replied:
@bennerv has assisted me in rewriting the unit test to test Reconcile rather than machineValid (same test cases as what was there previously). Let me know if this looks good.

m1kola reviewed (changes requested):
I think it is going in the right direction, but I spoted few things that need changing.

troy0820 commented:
Update the README with the new controller name and it's function.

mjudeikis commented on pkg/operator/controllers/machine/machine_controller.go:
```diff
@@ -0,0 +1,88 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	// "github.com/Azure/ARO-RP/pkg/operator"
```
Remove?

mjudeikis reviewed (changes requested):
I think its very close Just few nits and we can merge with follow-up story

m1kola commented on pkg/operator/controllers/machine/machine_controller.go:
```diff
@@ -0,0 +1,88 @@
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&arov1alpha1.Cluster{}, builder.WithPredicates(aroClusterPredicate)).
```
Bump https://github.com/Azure/ARO-RP/pull/1505#discussion_r645672041.

I do not see why we need to watch `arov1alpha1.Cluster{}`. If we are keeping it, could you please explain why?

cadenmarchese replied:
It's not needed, I've removed it.

m1kola reviewed (changes requested):
In general looks good, but there were few older conversations which got lost.

m1kola commented on pkg/operator/controllers/machine/machine_controller_test.go:
```diff
@@ -0,0 +1,322 @@
+
+			_, err := r.Reconcile(context.Background(), tt.request)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Machine.Reconcile() error = %v, wantErr %v", err, tt.wantErr)
+				return
```
1. `t.Fatal`/`t.Fatalf` is going to stop subtest execution.
2. We just tend to print an actual error. If the test fails - it will give you the test name + line number from which you can find the desired test result.

```suggestion
				t.Fatalf(err)
```

m1kola replied:
Also. It looks like you actually do not use `wantErr`. I think in this case it is ok not to test the error which `Reconcile` returns beucase the only error it returns comes from `controllers.SetCondition(...)` which should have a test own test. In master it currently has no test, but https://github.com/Azure/ARO-RP/pull/1508 add coverage.

So we can remove `wantErr`.

m1kola reviewed (changes requested):
The test looks good overall. I left few suggestions (nit picking really). The only real thing that stops me from merging this is incorrect `NewControllerManagedBy`.

I feel that this is going to be the last iteration :) Great work.

cadenmarchese commented on pkg/operator/controllers/machine/machine_controller_test.go:
```diff
@@ -0,0 +1,322 @@
+				t.Fatal("Error found")
+			}
+
+			if !cmp.Equal(cluster.Status.Conditions[0].Type, tt.wantConditions[0].Type) {
+				t.Errorf("Unexpected Type found\n want: %v\n got: %v", tt.wantConditions[0].Type, cluster.Status.Conditions[0].Type)
+			}
+
+			if !cmp.Equal(cluster.Status.Conditions[0].Status, tt.wantConditions[0].Status) {
+				t.Errorf("Unexpected Status found\n want: %v\n got: %v", tt.wantConditions[0].Status, cluster.Status.Conditions[0].Status)
+			}
+
+			if !cmp.Equal(strings.TrimSpace(cluster.Status.Conditions[0].Message), tt.wantConditions[0].Message) {
+				t.Errorf("Unexpected Message found\n want: %v\n got: %v", tt.wantConditions[0].Message, cluster.Status.Conditions[0].Message)
+			}
+
+			if !cmp.Equal(cluster.Status.Conditions[0].Reason, tt.wantConditions[0].Reason) {
+				t.Errorf("Unexpected Reason found\n want: %v\n got: %v", tt.wantConditions[0].Reason, cluster.Status.Conditions[0].Reason)
+			}
```
@m1kola and I spoke about this, but for others, the reason that there are multiple test assertions in this test rather than one is because `cluster.Status.Conditions` (which is vendored in) outputs `LastTransitionTime`, so tests always fail when attempting to compare both arrays because there's always a difference in timestamp. There's no simple way to exclude this, so we are leaving it as is (with simplified assertions) for the time being.

cadenmarchese commented on pkg/operator/controllers/machine/machine_controller.go:
```diff
@@ -0,0 +1,87 @@
+	})
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&machinev1beta1.Machine{}, builder.WithPredicates(aroClusterPredicate)).
```
I tested this suggestion and it works.

cadenmarchese commented on pkg/operator/controllers/machine/machine_controller_test.go:
```diff
@@ -0,0 +1,322 @@
+					{
+						Kind: "MachineSet",
+					},
+				},
```
This was left over - I have removed it.

m1kola commented on pkg/operator/controllers/machine/machine_controller_test.go:
```diff
@@ -0,0 +1,306 @@
+			_, err := r.Reconcile(context.Background(), tt.request)
+			if err != nil {
+				t.Fatal(err)
+				return
```
No need to explicitly `return` here: `t.Fatal` stops test/subtest execution.
```suggestion
```

m1kola commented on pkg/operator/controllers/machine/machine_controller_test.go:
```diff
@@ -0,0 +1,306 @@
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "foo1-hx8z7-master-1",
+				Namespace: machineSetsNamespace,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						Kind: "MachineSet",
+					},
+				},
```
Same as here: https://github.com/Azure/ARO-RP/pull/1505#discussion_r658669482. You said that these were leftovers. There are more in this file: if they are not important in the test - we should get rid of them.

m1kola commented on pkg/operator/controllers/machine/machine_controller_test.go:
```diff
@@ -0,0 +1,322 @@
+			}},
+		},
+		{
+			name:   "wrong_vm_size",
```
[This one](https://github.com/Azure/ARO-RP/pull/1505#discussion_r658665647) got missed. `"wrong vm size"`, `"wrong image publisher"`, `"wrong disk size"`, etc.

m1kola reviewed (commented):
@cadenmarchese I noticed that few small things got missed from the previous review round. Let's address them and it will be ready for a merge.

cadenmarchese commented:
Most recent push should have addressed all comments - let me know if there's anything else needed, or if I have missed something from the review history.

m1kola commented:
@mjudeikis and @nilsanderselde you previously requested changes here. Are you happy for this to get merged?

nilsanderselde commented on pkg/operator/controllers/machine/machine_controller_test.go:
```diff
@@ -0,0 +1,280 @@
+			},
+		}
+
+		worker2 := &machinev1beta1.Machine{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "foo-hx8z7-worker-2",
+				Namespace: machineSetsNamespace,
+				Labels:    map[string]string{"machine.openshift.io/cluster-api-machine-role": "worker"},
+			},
+			Spec: machinev1beta1.MachineSpec{
+				ProviderSpec: machinev1beta1.ProviderSpec{
+					Value: &runtime.RawExtension{
+						Raw: []byte(`{
+"apiVersion": "azureproviderconfig.openshift.io/v1beta1",
+"kind": "AzureMachineProviderSpec",
+"osDisk": {
+"diskSizeGB": 128
+},
+"image": {
+"publisher": "azureopenshift",
+"offer": "aro4"
+},
+"vmSize": "Standard_D4s_v3"
+}`)},
+				},
+			},
+		}
```
This differs from `worker1` only by name (`foo-hx8z7-worker-2`). To make this easier to read, can we do something like assign a `machinev1beta1.MachineSpec` variable above and use it twice, rather than repeating all this?

Same goes for the masters, of which all three are identical aside from the names. Let's just say it once.

mjudeikis replied:
https://github.com/Azure/ARO-RP/issues/1605

nilsanderselde reviewed (changes requested):
Looks good aside from some test boilerplate which should be boiled down.

mjudeikis commented:
@cadenmarchese please address @nilsanderselde comment in separate small nit PR! Thanks!
